File: 06. functions/testfunctions.py
"""
The following python program will  test the same functions as tested in
testfunctions.m

Note that the key difference between Matlab and Python is the following:

Matlab: Once a function is defined in a .m file and in the same folder structure,
the function may be called ad-hoc by simply its name and understanding that the
name is the same as the function name.

Python: Once a fuction is defined in a .py file and in the same folder structure,
the function must first be imported as shown below, referring to the * portion of
the *.py filename as the "library" you are importing from.  Once you have completed
this import, you can call the function by name.

Note: You have to make sure the file you are importing is in the same directory.

"""
from mymax import mymax

max1 = mymax(1, 5, 7)
print(max1)

# Note - there is no Workspace in Python.

"""
Annonymous functions are known as lambda functions in python.
"""

power = lambda x, n: x**n # define lambda function on the fly
print(power(2,4)) # call arguments that we defined above.

"""
Python Global Variables

"""

File: 06. functions/testfunctions.m
% The following .m file will run the function "mymax" which we created.
% This function simply finds the maximum of the three input numbers given.

[max] = mymax(1, 5, 7)

% note that any variables created while the function was running are not
% stored in the Workspace.

% Annonymous functions - you can create functions on the fly.
% f = @(arglist)expression
% The function itself will be stored in, "value" within the Workspace.

power = @(x, n) x.^n;
% We can call our above defined on-the-fly function and put them into a
% variable.  This function will remain stored in the variables.

result1 = power(7, 3)

% Matlab global variables
% Global variables can be declared which work inside all functions as shown:

global TOTAL;
TOTAL = 10;
n = [34, 45, 25, 45, 33, 19, 40, 34, 38, 42];
avg = average1(n)

% Other Notes on functions - functions can be "nested" meaning that functions can
% call other functions, mid-function.
